Extension/src/Debugger/lldb-dap.ts

```diff
@@ -0,0 +1,229 @@
+    /^lldb-vscode-\d+$^lldb-vscode-\d+\.exe$/
+];
+
+/** The search implementation to find the viable lldb-dap executable.
+ *
+ * We absolutely prefer an lldb-dap binary that is in the path.
+ * Failing that, we'll:
+ *  - try xcrun on OSX to see if xtools can give it to us
+ *  - try searching the secure well-known locations for a binary
+ *
+ * And if we can't find the actual lldb-dap binary, we'll try the same thing
+ * for 'lldb-vscode' (the old name for the DAP binary)
+ * and then 'lldb-dap-##' and 'lldb-vscode-##' - in practice, one might find a
+ * binary with the LLVM major version number in it.
+ *
+ * @returns The path to the lldb-dap executable or undefined if it was not found.
+ */
+async function searchForLldbDap() {
+    if (lldbDapPath) {
+        return lldbDapPath;
+    }
+    const start = Date.now();
+    // PATH binaries take priority.
+    for (const candidate of candidates) {
+        // First, search the environment PATH for the binary.
+        lldbDapPath = await findInPath(candidate, isValidLldbDap);
+        if (lldbDapPath) {
+            log(`Discovered lldb-dap binary for OSX at '${lldbDapPath}' ${Date.now() - start} msec`);
+            note(localize('lldb-dap.enabled', "The {0} debugger is enabled via '{1}'", 'lldb-dap', lldbDapPath));
+            return lldbDapPath;
+        }
+    }
+
+    // Well-known-locations next.
+    for (const candidate of candidates) {
+        if (isMacOS) {
+            // If that fails, use xcrun to find the path.
+            if (typeof candidate === 'string') {
+                lldbDapPath = await xcRun(candidate);
+                if (lldbDapPath) {
+                    log(`Discovered lldb-dap binary for OSX at '${lldbDapPath}' ${Date.now() - start} msec`);
+                    note(localize('lldb-dap.enabled', "The {0} debugger is enabled via '{1}'", 'lldb-dap', lldbDapPath));
+                    return lldbDapPath;
+                }
+            }
+
+            // If we got this far, it's not in the PATH or via xcrun - but worry not,
+            // we'll do a check of well known secured locations where it might be installed.
+            for (const folder of ['/Applications', '/opt/homebrew']) {
+                lldbDapPath = (await searchFolder(folder, candidate, isValidLldbDap, 8))[0];
+                if (lldbDapPath) {
+                    log(`Discovered lldb-dap binary for OSX at '${lldbDapPath}' ${Date.now() - start} msec`);
+                    note(localize('lldb-dap.enabled', "The {0} debugger is enabled via '{1}'", 'lldb-dap', lldbDapPath));
+                    return lldbDapPath;
+                }
+            }
+        }
+
+        if (isWindows) {
+            // If we got this far, it's not in the PATH - but worry not,
+            // we'll do a check of well known secured locations where it might be installed.
+            for (const folder of ['c:/Program Files/LLVM', 'c:/Program Files/', 'C:/program files (x86)/']) {

```

Comment:
Isn't c:/Program Files/LLVM redundant compared with c:/Program Files/?

Reply:
Not entirely - While yes, it overlaps, the idea is that we want to prioritize searching that folder first. The slight bit of overlap isn't going to be detrimental